--- edit	Tue Apr 23 23:28:45 2002
+++ edit.old	Sat May  4 00:55:36 2002
@@ -179,6 +184,10 @@
     $text =~ s/>/&gt\;/go;
     $text =~ s/\t/   /go;
 
+#AS added hook for plugins that want to do heavy stuff
+    &TWiki::Plugins::beforeEditHandler( $text, $topic, $webName );
+#/AS
+
     if( $TWiki::doLogTopicEdit ) {
         # write log entry
         &TWiki::Store::writeLog( "edit", "$webName.$topic", $extra );
