--- Plugins.pm	Sun May  5 09:43:41 2002
+++ Plugins.pm.old	Sun May  5 10:00:00 2002
@@ -18,12 +18,19 @@
 # This program applies in order the installed TWiki plugins.
 # Each plugin is a package that contains the subs:
 #
-#   initPlugin           ( $topic, $web, $user )
+#   initPlugin              ( $topic, $web, $user, $installWeb )
 #   commonTagsHandler    ( $text, $topic, $web )
-#   startRenderingHandler( $text, $web, $meta )
+#   startRenderingHandler   ( $text, $web )
 #   outsidePREHandler    ( $text )
 #   insidePREHandler     ( $text )
 #   endRenderingHandler  ( $text )
+#   beforeEditHandler       ( $text, $topic, $web )
+#   afterEditHandler        ( $text, $topic, $web )
+#   beforeSaveHandler       ( $text, $topic, $web )
+#   writeHeaderHandler      ( $query )
+#   redirectCgiQueryHandler ( $query, $url )
+#   getSessionValueHandler  ( $key )
+#   setSessionValueHandler  ( $key, $value )
 
 package TWiki::Plugins;
 
@@ -35,7 +42,11 @@
 	$VERSION
     );
 
-$VERSION = '1.000';
+# PLEASE!!! increase version if you add new Handlers!!!
+
+$VERSION  = '1.002'; # added Edit handlers
+#$VERSION = '1.001'; # added Save/Session/Header/Cgi handlers
+#$VERSION = '1.000'; # initial version
 
 @registrableHandlers = (
         'initPlugin',              # ( $topic, $web, $user, $installWeb )
@@ -44,6 +55,7 @@
         'outsidePREHandler',       # ( $text, $web )
         'insidePREHandler',        # ( $text, $web )
         'endRenderingHandler',     # ( $text, $topic, $web )
+        'beforeEditHandler',       # ( $text, $topic, $web )
         'afterEditHandler',        # ( $text, $topic, $web )
         'beforeSaveHandler',       # ( $text, $topic, $web )
         'writeHeaderHandler',      # ( $query )
@@ -354,6 +366,22 @@
     return &applyHandlers;
 }
 
+# =========================
+sub beforeEditHandler
+{
+    # Called by edit
+    unshift @_, ( 'beforeEditHandler' );
+    return &applyHandlers;
+}
+
+# =========================
+sub afterEditHandler
+{
+    # Called by edit
+    unshift @_, ( 'afterEditHandler' );
+    return &applyHandlers;
+}
+
 # =========================
 
 1;
