--- lib/TWiki/UI.pm	2004-05-29 08:51:35.000000000 +0200
+++ /home/micha/Projekt/twiki/cairo/lib/TWiki/UI.pm	2005-06-29 19:27:27.000000000 +0200
@@ -142,7 +142,7 @@
 
    return 1 if TWiki::Access::checkAccessPermission( $mode, $user, "",
                                                      $topic, $web );
-   oops( $webName, $topic, "access$mode" );
+   oops( $web, $topic, "access$mode" );
 
    return 0;
 }
