--- TWiki.cfg-old	Mon Oct 29 23:10:27 2001
+++ TWiki.cfg	Thu Oct 18 20:03:04 2001
@@ -142,7 +142,10 @@
 #                   RCS check in command with date :
 $revCiDateCmd     = "$rcsDir/ci -l $rcsArg -q -mnone -t-none -d$cmdQuote%DATE%$cmdQuote -w$cmdQuote%USERNAME%$cmdQuote %FILENAME%";
 #                   RCS check out command :
-$revCoCmd         = "$rcsDir/co $rcsArg -q -p%REVISION% %FILENAME%";
+# begin mtbear 2001-10-18
+$revCoCmd         = "$rcsDir/co $rcsArg -q -d%DATETIME% -p%REVISION% %FILENAME%";
+#$revCoCmd         = "$rcsDir/co $rcsArg -q -p%REVISION% %FILENAME%";
+# end mtbear 2001-10-18
 #                   RCS history command :
 $revHistCmd       = "$rcsDir/rlog $rcsArg -h %FILENAME%";
 #                   RCS history on revision command :
